Doc/library/__main__.rst

```diff
@@ -1,25 +1,182 @@
-
-:mod:`__main__` --- Top-level script environment
-================================================
+:mod:`__main__` --- CLIs, import-time behavior, and ``__name__ == '__main__'``
+==============================================================================
 
 .. module:: __main__
-   :synopsis: The environment where the top-level script is run.
+   :synopsis: Command line interfaces, import-time behavior, and ``__name__ == '__main__'``
 
 --------------
 
-``'__main__'`` is the name of the scope in which top-level code executes.
-A module's __name__ is set equal to ``'__main__'`` when read from
-standard input, a script, or from an interactive prompt.
+In Python, ``__main__`` is not a single mechanism in the language, but in fact
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
```

Comment:
Minor nit: `` __name__ == '__main__'`` is an expression right?  Whereas ``if __name__ == '__main__':`` is a statement.

Reply:
```suggestion
1. The ``if __name__ == '__main__':`` statement
```

Reply:
Good catch, but I am actually going to change 'statement' to 'expression', because the critical point is to explain the *expression* `__name__ == '__main__'`. The explanation about the idiom with an if statement and imports and all that needs to be held back so that we can address the expression **first.** Readers cannot understand the idioms that have grown around the expression without understanding why and how the expression evaluates in the first place.